src/main/java/au/com/southsky/jfreesane/SaneDevice.java

```diff
@@ -214,4 +217,41 @@ SaneSession getSession() {
   void invalidateOptions() {
     optionTitleMap = null;
   }
+
+  @Override
+  public int hashCode() {
+    int hash = 3;
+    hash = 37 * hash + Objects.hashCode(this.name);
+    hash = 37 * hash + Objects.hashCode(this.vendor);
+    hash = 37 * hash + Objects.hashCode(this.model);
+    hash = 37 * hash + Objects.hashCode(this.type);
+    return hash;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```

Comment:
Ok this helps my understanding! I think `SaneDevice` and `SaneOption` cannot readily be serialized, because they both have `SaneSession` fields. `SaneSession` in turn has a socket to the SANE server, so it can't be sensibly serialized.

One solution might be to introduce `saveOptionValues(OutputStream)` and `restoreOptionValues(InputStream)` methods on `SaneDevice`. Each time the user changes an option, we could add the option to a list of "dirty" options. When `saveOptionValues` is called, each dirty option is written to the output stream. `restore` will just call `getOption(...).set*Value` to restore the options.

All of this should probably be keyed by device name. What do you think?